File: server/src/main/kotlin/com/heerkirov/hedge/server/model/Tag.kt
package com.heerkirov.hedge.server.model

/**
 * 内容标签。
 */
data class Tag(val id: Int,
               /**
                * 排序下标，由系统维护，同一父标签一组从0开始。
                */
               val ordinal: Int,
               /**
                * 父标签的id。
                */
               val parentId: Int?,
               /**
                * 标签名。
                */
               val name: String,
               /**
                * 其他名称。
                */
               val otherNames: List<String>,
               /**
                * 标签类型。
                */
               val type: Type,
               /**
                * 组功能标记。
                * 组标记使一个地址段/标签的直接子节点被视作组员，建议只出现最多其一(添加二级子节点会推导出直接子节点，因此同样有效)。
                */
               val isGroup: IsGroup,
               /**
                * 描述。
                */
               val description: String,
               /**
                * 标签的颜色名称。
                */
               val color: String?,
               /**
                * 链接到其他标签。给出tag id列表。
                * 引入此标签时，链接到的其他标签会像此标签的父标签一样被导出。
                */
               val links: List<Int>?,
               /**
                * 标签的样例image。给出image id列表。
                */
               val examples: List<Int>?,
               /**
                * [exported field]关联的image的平均分。
                */
               val exportedScore: Int? = null,
               /**
                * [cache field]关联的image总数。仅包括image。
                */
               val cachedCount: Int = 0) {

    enum class Type {
        /**
         * 普通标签。
         */
        TAG,
        /**
         * 地址段。地址段不允许作为标签被引用，也不会出现在标签列表，但可以查询。
         */
        ADDR,
        /**
         * 虚拟地址段不会使用索引优化，因此查询效率有所损失，主要用于不需要索引的、用于分类归纳的地址段。
         */
        VIRTUAL_ADDR
    }

    enum class IsGroup {
        /**
         * 默认不开启。
         */
        NO,
        /**
         * 开启组标记。
         */
        YES,
        /**
         * 强制组：使建议升级为强制。但对推导无效。
         */
        FORCE,
        /**
         * 序列化组：使组员的ordinal排序具有实际意义，可以使用比较查询来查询组员。
         */
        SEQUENCE,
        /**
         * 强制+序列化组。
         */
        FORCE_AND_SEQUENCE
    }
}

File: frontend/src/functions/adapter-http/impl/tag.ts
import { HttpInstance, Response } from "../server"
import { IdResponse, LimitAndOffsetFilter, ListResult, OrderList } from "./generic"
import { SimpleIllust } from "./illust"

export function createTagEndpoint(http: HttpInstance): TagEndpoint {
    return {
        list: http.createQueryRequest("/api/tags"),
        tree: http.createQueryRequest("/api/tags/tree"),
        create: http.createDataRequest("/api/tags", "POST"),
        get: http.createPathRequest(id => `/api/tags/${id}`),
        update: http.createPathDataRequest(id => `/api/tags/${id}`, "PATCH"),
        delete: http.createPathRequest(id => `/api/tags/${id}`, "DELETE")
    }
}

/**
 * 标签。
 */
export interface TagEndpoint {
    /**
     * 查询标签列表。
     */
    list(filter: TagFilter): Promise<Response<ListResult<Tag>>>
    /**
     * 查询标签树。
     */
    tree(filter: TagTreeFilter): Promise<Response<TagTreeNode[]>>
    /**
     * 新建标签。
     * @exception NOT_EXIST ("parentId"|"links"|"examples"|"annotations", id) 指定的资源不存在
     * @exception NOT_SUITABLE ("examples"|"annotations", id) 指定的资源不适用。对于examples，只有类型为image的项目可用；对于annotations，是此注解的target要求不能应用于此种类的tag
     * @exception CANNOT_GIVE_COLOR 只有创建顶层标签时才能指定颜色
     * @exception ALREADY_EXISTS ("Tag", "name", name) 标签重名。addr类型的标签在同一个parent下禁止重名，tag类型的标签除上一条外还禁止与全局其他tag类型的标签重名
     */
    create(form: TagCreateForm): Promise<Response<IdResponse>>
    /**
     * 查看标签。
     * @exception NOT_FOUND
     */
    get(id: number): Promise<Response<DetailTag>>
    /**
     * 更改标签。
     * @exception NOT_FOUND
     * @exception NOT_EXIST ("parentId"|"links"|"examples"|"annotations", id) 指定的资源不存在
     * @exception NOT_SUITABLE ("examples"|"annotations|links", id) 指定的资源不适用。对于examples，只有类型为image的项目可用；对于annotations，是此注解的target要求不能应用于tag; 对于links，目标必须是非虚拟的
     * @exception RECURSIVE_PARENT 在父标签检查中发现了闭环
     * @exception CANNOT_GIVE_COLOR 只有顶层标签才能指定颜色
     * @exception ALREADY_EXISTS ("Tag", "name", name) 标签重名。addr类型的标签在同一个parent下禁止重名，tag类型的标签除上一条外还禁止与全局其他tag类型的标签重名
     */
    update(id: number, form: TagUpdateForm): Promise<Response<null>>
    /**
     * 删除标签。
     * @exception NOT_FOUND
     *
     */
    delete(id: number): Promise<Response<null>>
}

export type TagType = "TAG" | "ADDR" | "VIRTUAL_ADDR"
export type IsGroup = "NO" | "YES" | "FORCE" | "SEQUENCE" | "FORCE_AND_SEQUENCE"

export interface Tag {
    /**
     * tag id。
     */
    id: number
    /**
     * tag在当前parent下的排序顺位，从0开始。
     */
    ordinal: number
    /**
     * 父级tag的tag id。null表示没有父级tag，是顶层tag。
     */
    parentId: number | null
    /**
     * 标签名称。需要遵守tag name规范(不能为blank串，不能包含禁用字符' " ` . |)
     */
    name: string
    /**
     * 标签别名。需要遵守tag name规范(不能为blank串，不能包含禁用字符' " ` . |)，且长度不能超过32。
     */
    otherNames: string[]
    /**
     * 标签类型。
     */
    type: TagType
    /**
     * 标签的组属性。
     */
    group: IsGroup
    /**
     * 标签的颜色。颜色有与父标签同步的特性，因此只有根标签可以设置color，非根标签设置会抛出异常。
     */
    color: string | null
}

export interface DetailTag extends Tag {
    /**
     * 标签链接的其他标签的tag id。
     */
    links: number[]
    /**
     * 标签的描述内容。
     */
    description: string
    /**
     * 标签的样例。
     */
    examples: SimpleIllust[]
    /**
     * 标签的注解。
     */
    annotations: {id: number, name: string, canBeExported: boolean}[]
    /**
     * 标签关联的项目的平均分。
     */
    score: number | null
    /**
     * 标签关联的项目的数量。
     */
    count: number
}

export interface TagTreeNode {
    id: number
    name: string
    otherNames: string[]
    type: TagType
    group: IsGroup
    color: string | null
    children: TagTreeNode[] | null
}

export interface SimpleTag {
    id: number
    name: string
    color: string | null
    isExported: boolean
}

export interface TagCreateForm {
    name: string
    type: TagType
    parentId: number | null
    otherNames?: string[] | null
    ordinal?: number | null
    group?: IsGroup
    links?: number[]
    annotations?: (number | string)[] | null
    description?: string
    color?: string | null
    examples?: number[] | null
}

export interface TagUpdateForm {
    name?: string
    otherNames?: string[] | null
    ordinal?: number
    parentId?: number | null
    type?: TagType
    group?: IsGroup
    links?: number[] | null
    annotations?: (number | string)[] | null
    description?: string
    color?: string
    examples?: number[] | null
}

export interface TagFilter extends LimitAndOffsetFilter {
    search?: string | null
    order?: OrderList<"id" | "ordinal" | "name" | "createTime" | "updateTime">
    parent?: number
    type?: TagType
    group?: boolean
}

export interface TagTreeFilter {
    parent?: number
}
